pkg/placement/streams.go

```diff
@@ -0,0 +1,151 @@
+
+	"google.golang.org/grpc/metadata"
+
+	placementv1pb "github.com/dapr/dapr/pkg/proto/placement/v1"
+)
+
+type placementGRPCStream placementv1pb.Placement_ReportDaprStatusServer //nolint:nosnakecase
+
+type daprdStream struct {
+	id            uint32
+	hostName      string
+	hostID        string
+	hostNamespace string
+	needsVNodes   bool
+	stream        placementGRPCStream
+}
+
+func newDaprdStream(host *placementv1pb.Host, stream placementGRPCStream) *daprdStream {
+	return &daprdStream{
+		hostID:        host.GetId(),
+		hostName:      host.GetName(),
+		hostNamespace: host.GetNamespace(),
+		stream:        stream,
+		needsVNodes:   hostNeedsVNodes(stream),
+	}
+}
+
+// streamConnPool has the stream connections established between the placement gRPC server
+// and the Dapr runtime grouped by namespace, with an assigned id for faster lookup/deletion
+// The id is a simple auto-incrementing number, for efficiency.
+type streamConnPool struct {
+	lock sync.RWMutex // locks the streams map itself
+	// mutexMap *concurrency.MutexMap // manages locks on elements in the streams map
+
+	// Example representation of streams
+	//	{
+	//	  "ns1": {
+	//	  	1: stream1,
+	//	 	2: stream2,
+	//			3: stream3,
+	//		},
+	//
+	//	"ns2": {
+	//	  	4: stream4,
+	//	 	5: stream5,
+	//		},
+	//	}
+	streams map[string]map[uint32]*daprdStream
+
+	// streamIndexCnt assigns an index to streams in the streamConnPool.
+	// Its reset to zero every time a placement service loses leadership (thus clears all streams).
+	streamIndexCnt atomic.Uint32
+
+	// reverseLookup is a reverse index of streams to their daprdStream object.
+	// (so we don't have to loop through all streams to find a specific one)
+	reverseLookup map[placementGRPCStream]*daprdStream
+}
+
+func newStreamConnPool() *streamConnPool {
+	return &streamConnPool{
+		streams:       make(map[string]map[uint32]*daprdStream),
+		reverseLookup: make(map[placementGRPCStream]*daprdStream),
+	}
+}
+
+// add adds stream connection between runtime and placement to the namespaced dissemination pool.
+func (s *streamConnPool) add(stream *daprdStream) {
+	id := s.streamIndexCnt.Add(1)
+	stream.id = id
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if _, ok := s.streams[stream.hostNamespace]; !ok {
+		s.streams[stream.hostNamespace] = make(map[uint32]*daprdStream)
+	}
+
+	s.streams[stream.hostNamespace][id] = stream
+	s.reverseLookup[stream.stream] = stream
+}
+
+// delete removes a stream connection between runtime and placement
+// from the namespaced dissemination pool.
+// Returns true if the stream is the last one in a namespace
+func (s *streamConnPool) delete(stream *daprdStream) bool {
+	lastInNamespace := false
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if streams, ok := s.streams[stream.hostNamespace]; ok {
+		delete(streams, stream.id)
+		if len(streams) == 0 {
+			delete(s.streams, stream.hostNamespace)
+			delete(s.reverseLookup, stream.stream)
+			lastInNamespace = true
+		}
+	}
+
+	return lastInNamespace
```

Comment:
Where is this value used?

Reply:
Oh, you're right, it was a leftover from a refactor. Fixed.